<a href="https://colab.research.google.com/github/bidallei/MIAAD-UACJ/blob/main/AlvaroHernandez_SRFiltradoClase7_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

* Asignatura: Sistemas de recomendación
* Profesor: Dra. Karla Olmos Sánchez
* Alumno: **Álvaro Hernández Jarquín**
* Matrícula: 263150
* Práctica realizada: 09 de septiembre de 2026




In [1]:
import pandas as pd

ratings = pd.DataFrame({
    "usuario": [
        "Ana", "Ana", "Ana",
        "Bruno", "Bruno", "Bruno", "Bruno", "Bruno",
        "Maria", "Maria", "Maria", "Maria", "Maria",
        "Diego", "Diego", "Diego", "Diego", "Diego"
    ],
    "pelicula": [
        "Misión orbital", "Amor en otoño", "Risas en familia",
        "Misión orbital", "Amor en otoño", "Risas en familia",
        "Último dragón", "Casa silenciosa",
        "Misión orbital", "Amor en otoño", "Risas en familia",
        "Último dragón", "Casa silenciosa",
        "Misión orbital", "Amor en otoño", "Risas en familia",
        "Último dragón", "Casa silenciosa"
    ],
    "rating": [
        5, 2, 1,
        5, 1, 2, 5, 1,
        1, 5, 4, 1, 5,
        4, 2, 1, 5, 2
    ]
})

print (ratings)

matriz = ratings.pivot_table(
    index="usuario",
    columns="pelicula",
    values="rating"
)

print(matriz)

   usuario          pelicula  rating
0      Ana    Misión orbital       5
1      Ana     Amor en otoño       2
2      Ana  Risas en familia       1
3    Bruno    Misión orbital       5
4    Bruno     Amor en otoño       1
5    Bruno  Risas en familia       2
6    Bruno     Último dragón       5
7    Bruno   Casa silenciosa       1
8    Maria    Misión orbital       1
9    Maria     Amor en otoño       5
10   Maria  Risas en familia       4
11   Maria     Último dragón       1
12   Maria   Casa silenciosa       5
13   Diego    Misión orbital       4
14   Diego     Amor en otoño       2
15   Diego  Risas en familia       1
16   Diego     Último dragón       5
17   Diego   Casa silenciosa       2
pelicula  Amor en otoño  Casa silenciosa  Misión orbital  Risas en familia  \
usuario                                                                      
Ana                 2.0              NaN             5.0               1.0   
Bruno               1.0              1.0             5.0       

# Actividad 1

**Antes de ejecutar el código siguiente... investiga:**

¿Qué hace la instrucción pivot_table?
Pivotea la tabla del dataframe en base a una columna indice, lo que significa que convierte los valores categoricos de los renglones de una columna a columnas.

¿Qué tipo de dato es matriz?
Es un dataframe.

* No es necesario que escriban las respuestas


In [2]:
matriz = ratings.pivot_table(
    index="usuario",
    columns="pelicula",
    values="rating"
)

print(matriz)
type(matriz)

pelicula  Amor en otoño  Casa silenciosa  Misión orbital  Risas en familia  \
usuario                                                                      
Ana                 2.0              NaN             5.0               1.0   
Bruno               1.0              1.0             5.0               2.0   
Diego               2.0              2.0             4.0               1.0   
Maria               5.0              5.0             1.0               4.0   

pelicula  Último dragón  
usuario                  
Ana                 NaN  
Bruno               5.0  
Diego               5.0  
Maria               1.0  


pandas.core.frame.DataFrame

#Actividad 2

**Antes de ejecutar el siguiente código... Investiga:**

¿Qué hace la instrucción *loc*?
Selecciona las filas y columnas, en base a indices.

¿Qué contiene *datos*?
Dataframe, con lo filtrado.

¿Qué hace la instrucción *dropna*?
Quita los valores nulos, faltantes o None.

¿Qué tipo de dato es *comunes*?
Es un DataFrame.

Si se llamara la función de la siguiente forma:

  *ratings_comunes(matriz, Ana, Bruno)*

¿Cuál sería el resultado de *comunes*?
Regresaria los renglones y columnas de Ana y Bruno.


*   No es necesario escribir las respuestas




In [3]:
def ratings_comunes(matriz, usuario_1, usuario_2):
    datos = matriz.loc[[usuario_1, usuario_2]]

    print("datos", datos)
    comunes = datos.dropna(axis=1)
    print("comunes", comunes)
    return comunes

In [13]:
ratings_comunes(matriz, "Ana", "Bruno")

datos pelicula  Amor en otoño  Casa silenciosa  Misión orbital  Risas en familia  \
usuario                                                                      
Ana                 2.0              NaN             5.0               1.0   
Bruno               1.0              1.0             5.0               2.0   

pelicula  Último dragón  
usuario                  
Ana                 NaN  
Bruno               5.0  
comunes pelicula  Amor en otoño  Misión orbital  Risas en familia
usuario                                                  
Ana                 2.0             5.0               1.0
Bruno               1.0             5.0               2.0


pelicula,Amor en otoño,Misión orbital,Risas en familia
usuario,,,
Ana,2.0,5.0,1.0
Bruno,1.0,5.0,2.0


In [5]:
ratings_comunes(matriz, "Ana", "Maria")

datos pelicula  Amor en otoño  Casa silenciosa  Misión orbital  Risas en familia  \
usuario                                                                      
Ana                 2.0              NaN             5.0               1.0   
Maria               5.0              5.0             1.0               4.0   

pelicula  Último dragón  
usuario                  
Ana                 NaN  
Maria               1.0  
comunes pelicula  Amor en otoño  Misión orbital  Risas en familia
usuario                                                  
Ana                 2.0             5.0               1.0
Maria               5.0             1.0               4.0


pelicula,Amor en otoño,Misión orbital,Risas en familia
usuario,,,
Ana,2.0,5.0,1.0
Maria,5.0,1.0,4.0


In [6]:
ratings_comunes(matriz, "Ana", "Diego")

datos pelicula  Amor en otoño  Casa silenciosa  Misión orbital  Risas en familia  \
usuario                                                                      
Ana                 2.0              NaN             5.0               1.0   
Diego               2.0              2.0             4.0               1.0   

pelicula  Último dragón  
usuario                  
Ana                 NaN  
Diego               5.0  
comunes pelicula  Amor en otoño  Misión orbital  Risas en familia
usuario                                                  
Ana                 2.0             5.0               1.0
Diego               2.0             4.0               1.0


pelicula,Amor en otoño,Misión orbital,Risas en familia
usuario,,,
Ana,2.0,5.0,1.0
Diego,2.0,4.0,1.0


#Actividad 3

**Preguntas (Contesten en una celda de texto)**

¿Cuántos ítems tiene Ana en común con cada usuario?
Con Bruno 3.
Con Maria 3.
Con Diego 3.


¿Tener muchos ítems en común significa ser similar? ¿Por qué?

No, solo significa que calificaron las mismas películas, no que tengan valores parecidos.

¿Qué problema aparece si dos usuarios solamente coinciden en una película?
Que tenemos muy poca información respecto a la similitud, lo que daría un resultado muy pobre.

#Actividad 4

**Antes de ejecutar el siguiente código... Investiga:**

¿Por qué agregar la condición *if comunes.shape[1] == 0* ? Para quitar las películas que no calificaron, porque tienen valor 0

¿Qué valores tienen vector_1 y vector_2?
Son listas de Numpy, con las calificaciones de las películas de cada usuario.

¿Qué hace la instrucción *dot*?
Multiplica cada valor de las listas y luego los suma, es el producto punto.

¿Qué hace la instrucción *linalg.norm*?
Nos da como resultado la magnitud de un vector

In [7]:
import numpy as np

def similitud_coseno_usuarios(matriz, usuario_1, usuario_2):

    comunes = matriz.loc[[usuario_1, usuario_2]].dropna(axis=1)

    if comunes.shape[1] == 0:
        return None

    vector_1 = comunes.loc[usuario_1].values
    vector_2 = comunes.loc[usuario_2].values

    producto_punto = np.dot(vector_1, vector_2)

    norma_1 = np.linalg.norm(vector_1)
    norma_2 = np.linalg.norm(vector_2)

    similitud = producto_punto / (norma_1 * norma_2)

    return similitud

#Actividad 5

**Complementa la siguiente función:**

Se requiere calcular la similitud coseno para cada usuario con Ana

In [8]:
for usuario in matriz.index:

    if usuario != "Ana":

        similitud = similitud_coseno_usuarios(
            matriz,
            "Ana",
            usuario
        )

        print(usuario, similitud)

Bruno 0.9666666666666667
Diego 0.9960238411119947
Maria 0.5352643613280604


# Actividad independiente 4. Análisis de sensibilidad

Realiza los siguientes experimentos:

*Experimento A*

Cambiar el rating de Bruno para Misión orbital de 5 a 1.

*Experimento B*

Eliminar una calificación de Diego.

*Experimento C*

Agregar un usuario con una sola película en común con Ana.

*Experimento D*

Cambiar una calificación de Ana.

Comparar:

| Escenario                    | Similitud original | Similitud nueva | Cambio observado |
| ---------------------------- | -----------------: | --------------: | ---------------- |
| Bruno cambia un rating       |                    |                 |                  |
| Diego pierde un rating       |                    |                 |                  |
| Usuario con una coincidencia |                    |                 |                  |



**Preguntas**

¿Qué modificación produjo el mayor cambio?

¿La similitud es estable?

¿Confiaría en una similitud calculada con una sola coincidencia?

¿Qué condición mínima impondría para aceptar un vecino?

* Escribir en una celda de texto las respuestas




## **Experimento A**
Cambiar el rating de Bruno para Misión orbital de 5 a 1.

In [9]:
ratings_A = ratings.copy()

ratings_A.loc[
    (ratings_A["usuario"] == "Bruno") &
    (ratings_A["pelicula"] == "Misión orbital"),
    "rating"
] = 1

matriz_A = ratings_A.pivot_table(
    index="usuario",
    columns="pelicula",
    values="rating"
)

print(matriz_A)

sim_A = similitud_coseno_usuarios(matriz_A, "Ana", "Bruno")
print("Similitud Ana-Bruno (A):", sim_A)

pelicula  Amor en otoño  Casa silenciosa  Misión orbital  Risas en familia  \
usuario                                                                      
Ana                 2.0              NaN             5.0               1.0   
Bruno               1.0              1.0             1.0               2.0   
Diego               2.0              2.0             4.0               1.0   
Maria               5.0              5.0             1.0               4.0   

pelicula  Último dragón  
usuario                  
Ana                 NaN  
Bruno               5.0  
Diego               5.0  
Maria               1.0  
Similitud Ana-Bruno (A): 0.6708203932499369


## **Experimento B**

Eliminar una calificación de Diego.

In [10]:
ratings_B = ratings.copy()

ratings_B = ratings_B.drop(
    ratings_B[
        (ratings_B["usuario"] == "Diego") &
        (ratings_B["pelicula"] == "Misión orbital")
    ].index
)

matriz_B = ratings_B.pivot_table(
    index="usuario",
    columns="pelicula",
    values="rating"
)

print(matriz_B)

sim_B = similitud_coseno_usuarios(matriz_B, "Ana", "Diego")
print("Similitud Ana-Diego (B):", sim_B)

pelicula  Amor en otoño  Casa silenciosa  Misión orbital  Risas en familia  \
usuario                                                                      
Ana                 2.0              NaN             5.0               1.0   
Bruno               1.0              1.0             5.0               2.0   
Diego               2.0              2.0             NaN               1.0   
Maria               5.0              5.0             1.0               4.0   

pelicula  Último dragón  
usuario                  
Ana                 NaN  
Bruno               5.0  
Diego               5.0  
Maria               1.0  
Similitud Ana-Diego (B): 0.9999999999999998


## **Experimento C**

Agregar un usuario con una sola película en común con Ana.

In [11]:
ratings_C = ratings.copy()

nuevo_usuario = pd.DataFrame({
    "usuario": ["Pancho"],
    "pelicula": ["Misión orbital"],
    "rating": [1]
})

ratings_C = pd.concat([ratings_C, nuevo_usuario], ignore_index=True)

matriz_C = ratings_C.pivot_table(
    index="usuario",
    columns="pelicula",
    values="rating"
)

print(matriz_C)

sim_C = similitud_coseno_usuarios(matriz_C, "Ana", "Pancho")
print("Similitud Ana-Pancho (C):", sim_C)

pelicula  Amor en otoño  Casa silenciosa  Misión orbital  Risas en familia  \
usuario                                                                      
Ana                 2.0              NaN             5.0               1.0   
Bruno               1.0              1.0             5.0               2.0   
Diego               2.0              2.0             4.0               1.0   
Maria               5.0              5.0             1.0               4.0   
Pancho              NaN              NaN             1.0               NaN   

pelicula  Último dragón  
usuario                  
Ana                 NaN  
Bruno               5.0  
Diego               5.0  
Maria               1.0  
Pancho              NaN  
Similitud Ana-Pancho (C): 1.0


## **Experimento D**

Cambiar una calificación de Ana.

Elegimos cambiar “Misión orbital” de 5 a 3.

In [12]:
ratings_D = ratings.copy()

ratings_D.loc[
    (ratings_D["usuario"] == "Ana") & (ratings_D["pelicula"] == "Misión orbital"),
    "rating"
] = 3

matriz_D = ratings_D.pivot_table(
    index="usuario",
    columns="pelicula",
    values="rating"
)

print(matriz_D)

for usuario in ["Bruno", "Diego", "Maria"]:
    sim_D = similitud_coseno_usuarios(matriz_D, "Ana", usuario)
    print(f"Similitud Ana-{usuario} (D):", sim_D)

pelicula  Amor en otoño  Casa silenciosa  Misión orbital  Risas en familia  \
usuario                                                                      
Ana                 2.0              NaN             3.0               1.0   
Bruno               1.0              1.0             5.0               2.0   
Diego               2.0              2.0             4.0               1.0   
Maria               5.0              5.0             1.0               4.0   

pelicula  Último dragón  
usuario                  
Ana                 NaN  
Bruno               5.0  
Diego               5.0  
Maria               1.0  
Similitud Ana-Bruno (D): 0.9271050693011066
Similitud Ana-Diego (D): 0.9914601339836675
Similitud Ana-Maria (D): 0.7010681840159741


## Actividad independiente 4. Análisis de sensibilidad

| Escenario | Comparación | Similitud original | Similitud nueva | Cambio observado |
|---|---|---:|---:|---:|
| A. Bruno cambia rating *Misión orbital* 5→1 | Ana–Bruno | 0.9667 | 0.6708 | −0.2958 |
| B. Diego pierde rating *Misión orbital* | Ana–Diego | 0.9960 | 1.0000 | +0.0040 |
| C. Usuario con una sola coincidencia: Pancho solo *Misión orbital* | Ana–Pancho | No estaba | 1.0000 | Nuevo dato |
| D. Ana cambia rating *Misión orbital* 5→3 | Ana–Bruno | 0.9667 | 0.9271 | −0.0396 |
| D. Ana cambia rating *Misión orbital* 5→3 | Ana–Diego | 0.9960 | 0.9915 | −0.0046 |
| D. Ana cambia rating *Misión orbital* 5→3 | Ana–María | 0.5353 | 0.7011 | +0.1658 |

**¿Qué modificación produjo el mayor cambio?**

El mayor cambio se produjo en el Experimento A, con una variación absoluta de 0.2958 en la similitud de Ana y Bruno.
En el Experimento C no tenemos qué comparar porque no existía el usuario.
Si vemos el Experimento D, ahí se hizo el mayor cambio en el tercero, entre Ana y María, con un valor absoluto de 0.1658

**¿La similitud es estable?**

No es estable. Cuando cambiamos una calificación cambia, y más en este ejemplo que son pocos usuarios y películas.

**¿Confiaría en una similitud calculada con una sola coincidencia?**

No, cambia mucho la calificación aunque no tengan gustos parecidos, como el ejemplo C, que califican Ana y Pancho tan dispar y la similitud es 1.0.

**¿Qué condición mínima impondría para aceptar un vecino?**

Al menos 3 películas en común y que tengan una similitud mayor a 0.5, que es la mitad.